fs_burrito_store.js

```diff
@@ -1,64 +1,101 @@
-        if ("name" in idDetails) {
-            return idDetails["name"][lang];
-        } else {
-            return idDetails["id"];
-        }
+    super(configJson);
+    if (!fse.existsSync(sDir)) {
+      fse.mkdirSync(sDir, { recursive: false });
     }
+    this._metadataStore = new FSMetadataStore(this, sDir);
+    this._ingredientsStore = new FSIngredientsStore(this, sDir);
+    this._ingredientBuffer = new FSIngredientBuffer(this, sDir);
+  }
 
-    __metadataFromBundlePath(bundlePath) {
-        const metadataPath = path.join(bundlePath, "metadata.json");
-        return JSON.parse(fse.readFileSync(metadataPath));
+  idServerName(idServerId, nameLang) {
+    const lang = nameLang || 'en';
+    const idDetails = this._metadataStore._idServers[idServerId];
+    if ('name' in idDetails) {
+      return idDetails.name[lang];
     }
+    return idDetails.id;
+  }
 
-    __metadataFromDblBundlePath(bundlePath) {
-        const metadataPath = path.join(bundlePath, "metadata.xml");
-        const metadataString = fse.readFileSync(metadataPath, "utf-8");
-        const metadataDom = new xmldom.DOMParser().parseFromString(
-            metadataString,
-            "text/xml"
-        );
-        return new DBLImport(metadataDom).sbMetadata;
-    }
+  __metadataFromBundlePath(bundlePath) {
+    const metadataPath = path.join(bundlePath, 'metadata.json');
+    return JSON.parse(fse.readFileSync(metadataPath));
+  }
 
-    ingredientLocation(idServerId, entryId, revisionId, variantId, ingredientName) {
-        const md = this._metadataStore.__variantMetadata(idServerId, entryId, revisionId, variantId);
-        return this._ingredientsStore.__ingredientLocation(idServerId, entryId, revisionId, variantId, ingredientName, md);
-    }
+  __metadataFromDblBundlePath(bundlePath) {
+    const metadataPath = path.join(bundlePath, 'metadata.xml');
+    const metadataString = fse.readFileSync(metadataPath, 'utf-8');
+    const metadataDom = new xmldom.DOMParser().parseFromString(
+      metadataString,
+      'text/xml',
+    );
+    return new DBLImport(metadataDom).sbMetadata;
+  }
+
+  ingredientLocation(idServerId, entryId, revisionId, variantId, ingredientName) {
+    const md = this._metadataStore.__variantMetadata(idServerId, entryId, revisionId, variantId);
+    return this._ingredientsStore.__ingredientLocation(idServerId, entryId, revisionId, variantId, ingredientName, md);
+  }
 
+  exportToDir(idServerId, entryId, revisionId, variantId, toPath) {
+    const md = this._metadataStore.__variantMetadata(idServerId, entryId, revisionId, variantId);
+    if (!md) {
+      throw new BurritoError('VariantNotFoundInStore');
+    }
+    if (fse.existsSync(toPath)) {
+      throw new BurritoError('ExportPathAlreadyExists');
+    }
+    fse.mkdirSync(toPath, { recursive: false });
+    fse.writeFileSync(path.join(toPath, 'metadata.json'), JSON.stringify(md));
+    for (const [ingredientUrl, _] of Object.entries(
```

Comment:
I -think- this line can be reduced to `this.ingredients(...).entries().filter([x, y] => y)`, which can be reduced to `this.ingredients(...).values()`. Since you're throwing away the index, use values() over objects(), and each of these is part of the Object prototype, which `this` is certainly an instance of.

Reply:
and then I'd reduce that to a more functional style, like:
`this.ingredients(...).forEach(([ingredientUrl, ...]) => {  ... })`

Reply:
This is definitely the kind of review I'd like us to be doing.

However, in this case, I think you're misreading my code.

The filter gives us all the ingredients in the ingredients metadata (aka manifest) which are also present locally. y in this case is a boolean for "present locally". values() would fail if any metadata ingredient is not present locally (which shouldn't happen on idServers but commonly happens on clients).

The forEach won't work out of the box because, despite appearances to the contrary, "this" is not a pointer to the class instance. It's more like a pointer to "current scope" and forEach changes that, which means that you're no longer referring to the same "this" inside the forEach function. The most common idiom for dealing with this is

const self = this;

at the top of the method. I can work with this but, if we're going to do that, we really need to do it at the top of every single method because, otherwise, it becomes very hard to dip into code and understand what is happening.

for-of avoids this problem, which is one reason I like it - several lines less boilerplate, between the self declaration and the function signature. I also think that it's a good fit for cases like this where you're performing an action rather than returning a value. (Using "functional" syntax for side-effects isn't functional programming and, personally, I find that more confusing than an explicit loop.)

I would like to get rid of most of my C-style for loops though. The problem here is that not everything is iterable. XML nodelists are one painful example.

Reply:
My mistake